File: incident-platform-local-templates/templates/3-python-service/app/routers/health.py
from fastapi import APIRouter, Response, status
from datetime import datetime
import time
import psutil

from app.models import HealthCheck
from app.config import settings
from app.database import check_database_health

router = APIRouter()

# Track service start time
service_start_time = time.time()

@router.get("/health", response_model=HealthCheck)
async def health_check(response: Response):
    """
    Health check endpoint with dependency checks
    Returns 200 if healthy, 503 if degraded
    """
    health_status = "healthy"
    checks = {
        "database": "unknown",
        "memory": "healthy",
        "disk": "healthy"
    }
    
    # Check database
    if check_database_health():
        checks["database"] = "healthy"
    else:
        checks["database"] = "unhealthy"
        health_status = "degraded"
    
    # Check memory usage
    memory_percent = psutil.virtual_memory().percent
    if memory_percent > 90:
        checks["memory"] = "warning"
        health_status = "degraded"
    
    # Check disk usage
    disk_percent = psutil.disk_usage('/').percent
    if disk_percent > 90:
        checks["disk"] = "warning"
        health_status = "degraded"
    
    uptime = time.time() - service_start_time
    
    response.status_code = status.HTTP_200_OK if health_status == "healthy" else status.HTTP_503_SERVICE_UNAVAILABLE
    
    return HealthCheck(
        status=health_status,
        timestamp=datetime.utcnow(),
        service=settings.SERVICE_NAME,
        version="1.0.0",
        uptime=uptime,
        checks=checks
    )

@router.get("/health/ready")
async def readiness_check(response: Response):
    """Readiness probe - can service accept traffic?"""
    if check_database_health():
        return {"status": "ready"}
    else:
        response.status_code = status.HTTP_503_SERVICE_UNAVAILABLE
        return {"status": "not ready"}

@router.get("/health/live")
async def liveness_check():
    """Liveness probe - is service running?"""
    return {"status": "alive"}

File: incident-platform-local-templates/templates/2-nodejs-service/src/routes/health.js
const express = require('express');
const router = express.Router();
const { Pool } = require('pg');
const config = require('../config');

// Database connection pool
const pool = new Pool({
  connectionString: config.database.url,
  max: 20,
  idleTimeoutMillis: 30000,
  connectionTimeoutMillis: 2000,
});

// Health check endpoint
router.get('/', async (req, res) => {
  const healthCheck = {
    status: 'healthy',
    timestamp: new Date().toISOString(),
    service: 'alert-ingestion-service',
    version: '1.0.0',
    uptime: process.uptime(),
    checks: {
      database: 'unknown',
      memory: 'healthy'
    }
  };

  try {
    // Check database connection
    const dbCheck = await pool.query('SELECT NOW()');
    healthCheck.checks.database = 'healthy';
  } catch (error) {
    healthCheck.checks.database = 'unhealthy';
    healthCheck.status = 'degraded';
    console.error('Database health check failed:', error);
  }

  // Check memory usage
  const memUsage = process.memoryUsage();
  const memUsagePercent = (memUsage.heapUsed / memUsage.heapTotal) * 100;
  if (memUsagePercent > 90) {
    healthCheck.checks.memory = 'warning';
    healthCheck.status = 'degraded';
  }

  const statusCode = healthCheck.status === 'healthy' ? 200 : 503;
  res.status(statusCode).json(healthCheck);
});

// Readiness check
router.get('/ready', async (req, res) => {
  try {
    await pool.query('SELECT 1');
    res.status(200).json({ status: 'ready' });
  } catch (error) {
    res.status(503).json({ status: 'not ready', error: error.message });
  }
});

// Liveness check
router.get('/live', (req, res) => {
  res.status(200).json({ status: 'alive' });
});

module.exports = router;
